main/server.go

```diff
@@ -0,0 +1,313 @@
+}
+
+// GetTodo handles GET requsts to /todos/{todoID}.
+// There's no parameters and it returns a JSON encoded todo.
+//
+// Examples:
+//
+//   req: GET /todos/1
+//   res: 200 {"id": 1, "title": "Buy bread", "completed": true}
+//
+//   req: GET /todos/42
+//   res: 404 todo not found
+func GetTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	t, ok := todo.Get(appengine.NewContext(r), id)
+	if !ok {
+		return notFound{}
+	}
+	return writeJson(w, http.StatusOK, t)
+}
+
+// UpdateTodo handles PUT requests to /todos/{todoID}.
+// The request body must contain a JSON encoded todo.
+// The id property is optional, but if it is included, it must match
+// with the request path.
+//
+// Example:
+//
+//   req: PUT /todos/1 {"id": 1, "title": "Learn Go", "completed": true}
+//   res: 200
+//
+//   req: PUT /todos/2 {"id": 1, "title": "Learn Go", "completed": true}
+//   res: 400 inconsistent todo IDs
+func UpdateTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	var t todo.Todo
+	if err := json.NewDecoder(r.Body).Decode(&t); err != nil {
+		return badRequest{err}
+	}
+	if t.ID == 0 {
+		t.ID = id
+	}
+	if t.ID != id {
+		return badRequest{fmt.Errorf("inconsistent todo IDs")}
+	}
+	if _, ok := todo.Get(c, id); !ok {
+		log.Infof(c, "Unable to find todo: %v", t)
+		return notFound{}
+	}
+	if err = t.Save(c); err != nil {
+		return err
+	}
+	return writeJson(w, http.StatusOK, t)
+}
+
+// DeleteTodo handles DELETE requests to /todos/{todoID}.
+// Returns a badRequest error if the ID cannot be parsed, and notFound if
+// no corresponding todo can be found.
+func DeleteTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	log.Infof(c, "Trying to delete id %v", id)
+	if ok := todo.Delete(c, id); !ok {
+		return notFound{}
+	}
+	w.WriteHeader(http.StatusNoContent)
+	return nil
+}
+
+// DeleteCompletedTodos handles DELETE requests to /todos.
+// It attempts to delete all todos which have been marked completed and returns
+// an error if one occurred.
+func DeleteCompletedTodos(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	if err := todo.DeleteCompleted(c); err != nil {
+		return err
+	}
+	w.WriteHeader(http.StatusNoContent)
+	return nil
+}
+
+// mergeJsonObjectFiles merges the k/v pairs in all indicated files with the
+// base map derived from the first file. An error is returned if any file
+// cannot be read or have its data unmarshaled. An error is also returned
+// if the resulting map cannot be marshaled.
+// Importantly, this function only works with files corresponding to JSON
+// objects! Files that correspond to JSON arrays will fail!
+func mergeJsonObjectFiles(files ...string) (b []byte, err error) {
+	if files == nil {
```

Comment:
Really? As per the [spec](https://golang.org/ref/spec#Passing_arguments_to_..._parameters): "If f is variadic with a final parameter p of type ...T, then within f the type of p is equivalent to type []T. If f is invoked with no actual arguments for p, the value passed to p is nil."